File: src/positioning/keti_ws/test/ThreeAnchPos3.py
import numpy as np

def ThreeAnchPos3(xa, ya, dist):
    M = len(xa)
    A = np.zeros((M-1, 2))
    y = np.zeros((M-1, 1))

    for k in range(M-1):
        A[k, :] = [-2 * (xa[k] - xa[M-1]), -2 * (ya[k] - ya[M-1])]
        y[k, 1] = dist[k]**2 - dist[M-1]**2 - (xa[k]**2 - xa[M-1]**2) - (ya[k]**2 - ya[M-1]**2)

    Pos = np.linalg.inv(A.T @ A) @ A.T @ y
    Prob = np.sqrt(np.mean(np.abs((xa - Pos[0])**2 + (ya - Pos[1])**2 - dist**2)))

    return Pos, Prob

File: src/positioning/matlab/MATLAB 4/ThreeAnchPos3.m
function [Pos,Prob]=ThreeAnchPos3(xa, ya, dist)
M = length(xa);
A = zeros(M-1,2);
y = zeros(M-1,1);


% if (ya(2)-y(1))*(xa(3)-xa(1)) ~= (ya(3)-y(1))*(xa(2)-xa(1))
    for k = 1 : M-1 
        A(k,:) = [-2*(xa(k)-xa(M)) -2*(ya(k)-ya(M))];
        y(k,1) = dist(k)^2-dist(M)^2-(xa(k)^2-xa(M)^2)-(ya(k)^2-ya(M)^2);
    end
% else
%     
% end


% Pos = (A'*A)\(A'*y);
Pos = inv(A'*A)*A'*y;

Prob = sqrt(mean(abs((xa - Pos(1)).^2 + (ya - Pos(2)).^2 - dist.^2)));

                      